### PR TITLE
Added UnsupportedTypeError and QError

### PR DESCRIPTION
**Pull request by jkessler93**
Added UnsupportedTypeError and QError to decode.go

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -421,7 +422,7 @@ func readData(r *bufio.Reader, order binary.ByteOrder) (*K, error) {
 			return nil, err
 		}
 		errmsg := string(line[:len(line)-1])
-		return nil, errors.New(errmsg)
+		return nil, NewQError(errmsg)
 	}
-	return nil, ErrBadMsg
+	return nil, NewUnsupportedTypeError(fmt.Sprintf("Unsupported type: %v", msgtype))
```
Comment (yznima) [inline decode.go]: Why is this needed?

File: struct.go [hunk 2]
```diff
@@ -329,15 +329,40 @@ func (k K) String() string {
...
 
 // Cannot process sync requests
 var ErrSyncRequest = errors.New("nosyncrequest")
 
+// NewQError initializes a new QError with a message.
+func NewUnsupportedTypeError(msg string) error {
+	return &UnsupportedTypeError{msg}
+}
+
+// UnsupportedTypeError is an error returned by Fn() calls. It can be used to distinguish these errors from generic errors
```
Comment (yznima) [inline struct.go]: (minor) `Fn()` is a concept in Appian gateways. I think it will be confusing mentioning it here. I think it will be better to say something along the lines of

`error returned when encoding or decoding a type that is not supported by kdb`

File: struct.go [hunk 3]
```diff
@@ -329,15 +329,40 @@ func (k K) String() string {
...
+	return e.msg
+}
+
+// NewQError initializes a new QError with a message.
+func NewQError(msg string) error {
+	return &QError{msg}
+}
+
+// QError is an error returned by Fn() calls. It can be used to distinguish these errors from generic errors
+type QError struct {
```
Comment (yznima) [inline struct.go]: Could we rename this to something that doesn't have `Q` in it? My reasoning is that q is the language and not the actual runner. So in a sense, there is no `GoError` or `JavaError`. I think another alternative is `Error`(`kdb.Error`), or `RuntimeError`(`kdb.RuntimeError`)

File: decode.go (hunk 1 above)
Comment (yznima) [inline decode.go]: `%d`?

Comment (yznima) [reply decode.go]: Also what about line `418`?

File: struct.go [hunk 4]
```diff
@@ -329,15 +329,40 @@ func (k K) String() string {
...
-// Message is malformated or invalid
-var ErrBadMsg = errors.New("Bad Message")
-
 // Message header is invalid
 var ErrBadHeader = errors.New("Bad header")
 
 // Cannot process sync requests
 var ErrSyncRequest = errors.New("nosyncrequest")
 
+// NewQError initializes a new QError with a message.
```
Comment (sbenderli) [inline struct.go]: comment should be `NewUnsupportedTypeError`

**Comment by yznima** [comment]
You could run the encode & decode test as 

```
go test --short -v --run TestEn
go test --short -v --run TestDe
```

**Comment by jkessler93** [comment]
@appian/datalayer-ha ready for another look

**Review by yznima — APPROVED** [review]
💯

**Review by sbenderli — APPROVED** [review]
LGTM
